Add Annoy as VectorStore

bakebrain opened the pull request:
Adds Annoy (https://github.com/spotify/annoy) as vector Store. 

RESOLVES hwchase17/langchain#2842

discord ref:  https://discord.com/channels/1038097195422978059/1051632794427723827/1096089994168377354

bakebrain committed:
add Annoy as VectorStore

- basic functionality  - closely related to the FAISS implementation
- added some tests

bakebrain committed:
adjusted docstrings and added some more params

bakebrain committed:
Add Annoy documentation

bakebrain committed:
fixing lint issues

bakebrain committed:
fix typos and links in docs

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        n_jobs: int = -1,
+        **kwargs: Any,
+    ) -> Annoy:
+        assert metric in INDEX_METRICS, f"Invalid metric {metric}"
```
```suggestion
        if metric not in INDEX_METRICS:
            raise ValueError(f"Unsupported distance metric: {metric}. Expected one of {list(INDEX_METRICS)}")
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+    ) -> Annoy:
+        assert metric in INDEX_METRICS, f"Invalid metric {metric}"
+        annoy = dependable_annoy_import()
+        f = len(embeddings[0])
```
```suggestion
        if not embeddings:
            raise ValueError("embeddings must be provided to build AnnoyIndex")
        f = len(embeddings[0])
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+            f: Number of dimensions in the embeddings.
+            metric: Metric to use for the index.
+        """
+        assert metric in INDEX_METRICS, f"Invalid metric {metric}"
```
```suggestion
        if metric not in INDEX_METRICS:
            raise ValueError(f"Unsupported distance metric: {metric}. Expected one of {list(INDEX_METRICS)}")
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        cls,
+        folder_path: str,
+        embeddings: Embeddings,
+        f: int = 0,
```
```
        f: Optional[int] = None,
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        """
+        path = Path(folder_path)
+        os.makedirs(path, exist_ok=True)
+        # save index
```
```suggestion
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        os.makedirs(path, exist_ok=True)
+        # save index
+        self.index.save(str(path / "index.annoy"), prefault=prefault)
+        # save docstore and index_to_docstore_id
```
```suggestion
```

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        if not f:
+            f = len(embeddings.embed_documents(["dimension"])[0])
+
+        index = annoy.AnnoyIndex(f, metric=metric)
```
If the index was saved with a certain metric and then loaded with a different metric, that would create an error, right?

bakebrain replied:
Yes - it will fail

vowelparrot commented on langchain/vectorstores/annoy.py:
```diff
@@ -0,0 +1,414 @@
+        annoy = dependable_annoy_import()
+        # derive embedding dimensionality from random doc if not provided
+        if not f:
+            f = len(embeddings.embed_documents(["dimension"])[0])
```
Rather than re-compute the embedding dimension, can we save a `config.json` in the index directory with the metric and embedding dimension? This vector store is read-only anyway

bakebrain replied:
yes, indeed: storing the config is the obvious way to go - with that,  `load_local` relies only on the path and the embedding method. I used configparser and added it to the pickle

vowelparrot reviewed (commented):
Thank you for the PR!

hwchase17 committed:
Update langchain/vectorstores/annoy.py

Co-authored-by: vowelparrot <130414180+vowelparrot@users.noreply.github.com>

hwchase17 committed:
Update langchain/vectorstores/annoy.py

Co-authored-by: vowelparrot <130414180+vowelparrot@users.noreply.github.com>

hwchase17 committed:
Update langchain/vectorstores/annoy.py

Co-authored-by: vowelparrot <130414180+vowelparrot@users.noreply.github.com>

bakebrain committed:
add index config

adding index config to the pickle which allows to not derive the embedding dimension explicitly. Annoy does not allow to read the metric from the index itself - adding it explicitly to the class

bakebrain committed:
cleanups

bakebrain commented:
Thanks for the comments - I store the "index config" now explicitly as part of the pickle via `ConfigParser`. Unfortunately one cannot get the used metric from the index itself (the dimensionality is possible though) - so I just added it explicitly. Still better like so imho.

bakebrain committed:
Merge branch 'hwchase17:master' into annoy

bakebrain committed:
adjust to changes in base class

bakebrain committed:
adding missing elasticsearch doc file

hwchase17 reviewed (approved):
this is awesome! thanks :)
